### PR TITLE
[FEAT]: Babybear for golang

### PR DESCRIPTION
**Pull request by jeremyfelder**
## Describe the changes

This PR:
- Moves all curve/field agnostic checks to `core` package
- Adds babybear field
  - Changed field representation back to using 32 bit integers
- Adds grumpkin curve
- Changes the generation and structure of the golang bindings
  - There is a configuration file for each field and curve
  - Each additional functionality of the API is generated on its own per field or curve and is a separate package within the field or curve package. This allows integrations to only compile what they need and not all of our supported primitives.
- Generating files can be done using `go generate ./...` from any directory within the project

### CODE REVIEW

File: .github/workflows/golang.yml [hunk 1]
```diff
@@ -33,13 +33,23 @@ jobs:
...
+          - name: bn254
+            build_args: -g2 -ecntt
+          - name: bls12_381
+            build_args: -g2 -ecntt
+          - name: bls12_377
+            build_args: -g2 -ecntt
+          - name: bw6_761
+            build_args: -g2 -ecntt
+          - name: grumpkin
+            build_args:
```
Comment (ChickenLover) [inline .github/workflows/golang.yml:52]: TODO Roman: implement the same for our c++ workflow (rn it just doesn't test grumpkin at all)

File: wrappers/golang/build.sh [hunk 2]
```diff
@@ -57,4 +91,17 @@ do
...
+# Needs to remove the CMakeCache.txt file to allow building fields after curves
+# have been built since CURVE and FIELD cannot both be defined
+rm -f "$BUILD_DIR/CMakeCache.txt"
+
+for FIELD in "${BUILD_FIELDS[@]}"
+do
+  echo "FIELD=${FIELD}" > build_config.txt
+  echo "DEVMODE=${DEVMODE}" >> build_config.txt
+  cmake -DCMAKE_CUDA_COMPILER=$CUDA_COMPILER_PATH -DFIELD=$FIELD -DEXT_FIELD=$EXT_FIELD -DDEVMODE=$DEVMODE -DCMAKE_BUILD_TYPE=Release -S . -B build
+  cmake --build build -j8 && rm build_config.txt
```
Comment (ChickenLover) [inline wrappers/golang/build.sh:105]: `-j8` => `-j`

File: wrappers/golang/internal/generator/ecntt/templates/ecntt.h.tmpl [hunk 3]
```diff
@@ -1,5 +1,4 @@
 #include <cuda_runtime.h>
-#include "../../../include/types.h"
```
Comment (ChickenLover) [inline wrappers/golang/internal/generator/ecntt/templates/ecntt.h.tmpl:2]: how go knows about NTTConfig and projective_t if there is nothing else included?

File: wrappers/golang/internal/generator/lib_linker/generate.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package libe_linker
```
Comment (ChickenLover) [inline wrappers/golang/internal/generator/lib_linker/generate.go:1]: typo?

**Review by ChickenLover — APPROVED** [review]
Mostly reviewed the c++ part and the new code gen script. Configs looks nice.

File: wrappers/golang/internal/generator/ecntt/templates/ecntt.h.tmpl (hunk 3 above)
Comment (jeremyfelder) [reply wrappers/golang/internal/generator/ecntt/templates/ecntt.h.tmpl:2]: They're partially declared lower in this file: https://github.com/ingonyama-zk/icicle/pull/481/files/2043ed13721067aba9e71ee381a41fd402a8958d#diff-ebd41d829ee4e442d0136366489b1f9d02faf670ac9553cdf7b7f21386bde13fR10-R11

File: wrappers/golang/internal/generator/lib_linker/generate.go (hunk 4 above)
Comment (jeremyfelder) [reply wrappers/golang/internal/generator/lib_linker/generate.go:1]: Yes, nice catch 👍🏻

File: wrappers/golang/build.sh (hunk 2 above)
Comment (jeremyfelder) [reply wrappers/golang/build.sh:105]: Why do we want to use the build tool's default number of jobs?

Comment (LeonHibnik) [reply wrappers/golang/build.sh:105]: What's the default number here?
Maybe it's better to set it to use all the cpu cores

Comment (ChickenLover) [reply wrappers/golang/build.sh:105]: I actually thought that -j will use num-cpus threads

Comment (jeremyfelder) [reply wrappers/golang/build.sh:105]: https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-build-j. Lets leave as is for now unless we have a reason to change it

**Pull request merged**